src/json/builder.cr

```diff
@@ -94,35 +94,55 @@ class JSON::Builder
   # This method can also be used to write the name of an object field.
   def string(value)
     string = value.to_s
+
     scalar(string: true) do
       io << '"'
-      string.each_char do |char|
-        case char
+
+      pos = 0
+      reader = Char::Reader.new(string)
+
+      while reader.has_next?
+        set_pos = true
+        escape = nil
+
+        case char = reader.current_char
         when '\\'
-          io << "\\\\"
+          escape = "\\\\"
         when '"'
-          io << "\\\""
+          escape = "\\\""
         when '\b'
-          io << "\\b"
+          escape = "\\b"
         when '\f'
-          io << "\\f"
+          escape = "\\f"
         when '\n'
-          io << "\\n"
+          escape = "\\n"
         when '\r'
-          io << "\\r"
+          escape = "\\r"
         when '\t'
-          io << "\\t"
+          escape = "\\t"
         when .ascii_control?
+          io.write string.to_slice[pos, reader.pos - pos]
           io << "\\u"
           ord = char.ord
           io << '0' if ord < 0x1000
           io << '0' if ord < 0x100
           io << '0' if ord < 0x10
           ord.to_s(16, io)
         else
-          io << char
+          set_pos = false
         end
+
+        if escape
```

Comment:
I'd suggest to put this behaviour directly in the `when` branches (maybe extract in a private method). This adds some code duplication but eliminates an additional condition check.

Or just skip to the next char with `next` from the `.ascii_control?` and `else` branches. Then you can keep it here but don't need the condition.

Reply:
put directly in `when`: longer code.

i understand your concern, but these little details not improve performance, and make code longer, same to understand. so i prefer not change.